src/transformers/models/segformer/modeling_tf_segformer.py

```diff
@@ -0,0 +1,878 @@
+                height = tf.math.sqrt(tf.cast(shape_list(encoder_hidden_state)[1], tf.float32))
+                height = width = tf.cast(height, tf.int32)
+                encoder_hidden_state = tf.reshape(encoder_hidden_state, (batch_size, height, width, -1))
+
+            # unify channel dimension
+            encoder_hidden_state = tf.transpose(encoder_hidden_state, perm=[0, 2, 3, 1])
+            height = shape_list(encoder_hidden_state)[1]
+            width = shape_list(encoder_hidden_state)[2]
+            encoder_hidden_state = mlp(encoder_hidden_state)
+            encoder_hidden_state = tf.reshape(encoder_hidden_state, (batch_size, height, width, -1))
+
+            # upsample
+            temp_state = tf.transpose(encoder_hidden_states[0], perm=[0, 2, 3, 1])
+            upsample_resolution = shape_list(temp_state)[1:-1]
+            encoder_hidden_state = tf.image.resize(encoder_hidden_state, size=upsample_resolution, method="bilinear")
+            all_hidden_states += (encoder_hidden_state,)
+
+        hidden_states = self.linear_fuse(tf.concat(all_hidden_states[::-1], axis=-1))
+        hidden_states = self.batch_norm(hidden_states)
+        hidden_states = self.activation(hidden_states)
+        hidden_states = self.dropout(hidden_states)
+
+        # logits of shape (batch_size, height/4, width/4, num_labels)
+        logits = self.classifier(hidden_states)
+
+        return logits
+
+
+@add_start_docstrings(
+    """SegFormer Model transformer with an all-MLP decode head on top e.g. for ADE20k, CityScapes.""",
+    SEGFORMER_START_DOCSTRING,
+)
+class TFSegformerForSemanticSegmentation(TFSegformerPreTrainedModel):
+    def __init__(self, config: SegformerConfig, **kwargs):
+        super().__init__(config, **kwargs)
+        self.segformer = TFSegformerMainLayer(config, name="segformer")
+        self.decode_head = TFSegformerDecodeHead(config, name="decode_head")
+
+    def hf_compute_loss(self, logits, labels):
+        # upsample logits to the images' original size
+        if len(shape_list(labels)) > 3:
+            label_interp_shape = shape_list(labels)[1:-1]
+        else:
+            label_interp_shape = shape_list(labels)[-2:]
+
+        upsampled_logits = tf.image.resize(logits, size=label_interp_shape, method="bilinear")
+        # compute weighted loss
+        loss_fct = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction="none")
+
+        # Adapted from https://www.tensorflow.org/text/tutorials/transformer#loss_and_metrics.
+        # Utility to mask the index to ignore during computing the loss.
+        def masked_loss(real, pred):
+            mask = tf.math.logical_not(tf.math.equal(real, self.config.semantic_loss_ignore_index))
+            loss_ = loss_fct(real, pred)
+            mask = tf.cast(mask, dtype=loss_.dtype)
+            loss_ *= mask
+            return loss_  # No reduction since other HF losses don't do it.
```

Comment:
@Rocketknight1 I have kept the loss non-reduced. This means the loss will have a shape of `(batch_size, height, width)`. For this, I also had to change the loss computation test accordingly (see [here](https://github.com/sayakpaul/transformers/blob/8d913e6bae8b9f5d8c857c44f664aad3bdc8205e/tests/models/segformer/test_modeling_tf_segformer.py#L418-#L424)). Curious to know your thoughts. 

Note that the PT variant applies the reduction:

https://github.com/huggingface/transformers/blob/692e61e91a0b83f5b847902ed619b7c74c0a5dda/src/transformers/models/segformer/modeling_segformer.py#L814

But since `hf_compute_loss()` doesn't apply any kind of reduction, I thought of keeping this one like so.

Reply:
This is actually the thing I've been working on! Our TensorFlow losses were out-of-date; they used methods like `tf.boolean_mask()` and therefore were not compatible with XLA. 

I think I would prefer to standardize loss shapes like we do in our PyTorch models. My preference would be to return a loss vector of shape `(batch_size,)`, because this is what Keras expects, and that means users can use `sample_weights` if they want to. However, just returning a scalar loss is fine for most purposes too.

In this case, I think it's fine to reduce to either a 1D vector of per-sample losses or a single loss scalar. We might revisit that when we decide on our official policy for TF losses.

Reply:
I am okay with either of those two too. But here's my only question: https://github.com/huggingface/transformers/pull/17910#discussion_r912310915